### PR TITLE
#63 Using embedded etcd server as a payment channel storage

### PR DESCRIPTION
**Pull request by stellarspot**
This is an initial fix which includes etcd as a payment storage into snet-daemon project.
It includes two main structures 
* EtcdServer
* EtcdClient
and some basic operations on it.

The etcddb_test.go test show basic usage like using a config file to start an etcd server and read/write values to it.

The example of the config file:
```
{
	"PAYMENT_CHANNEL_STORAGE_CLUSTER": "storage_1=http://127.0.0.1:2380",

	"PAYMENT_CHANNEL_STORAGE": {
		"ID": "storage_1",
		"HOST" : "127.0.0.1",			
		"CLIENT_PORT": 2379,
		"PEER_PORT": 2380,
		"TOKEN": "payment_channel_storage_token",
		"ENABLED": true
	}
}
```

### CODE REVIEW

File: etcddb/payment_channel_storage.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+)
+
+// PaymentChannelStorageConf contains embedded etcd server conf
+type PaymentChannelStorageConf struct {
+	ID         string
+	Scheme     string
+	Host       string
+	ClientPort int `mapstructure:"CLIENT_PORT"`
+	PeerPort   int `mapstructure:"PEER_PORT"`
+	Token      string
```
[the remark this answers is not in the document]
Comment (stellarspot) [reply etcddb/payment_channel_storage.go]: I updated the code to use token in the etcd server.

File: etcddb/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+Note that it is possible to disable running an embedded etcd server in the config file by setting *ENABLED* field to false:
+
+```json
+{
+    "PAYMENT_CHANNEL_STORAGE_CLUSTER": "storage_1=http://127.0.0.1:2380,storage_2=http://127.0.0.2:2380,storage_3=http://127.0.0.3:2380",
+
+    "PAYMENT_CHANNEL_STORAGE": {
+        "ID": "storage_2",
+        "HOST" : "127.0.0.2",
```
[the remark this answers is not in the document]
Comment (stellarspot) [reply etcddb/README.md]: Updated.

File: etcddb/etcddb_test.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	assert.Equal(t, "payment_channel_storage_token", conf.Token)
+	assert.Equal(t, true, conf.Enabled)
+
+	server, err := InitEtcdServer(vip)
+
+	assert.Nil(t, err)
+	assert.NotNil(t, server)
+	defer server.Close()
+}
+func TestPaymentChannelStorageReadWrite(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (stellarspot) [reply etcddb/etcddb_test.go]: CAS test is added.

File: etcddb/etcddb_test.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
+package etcddb
```
[the remark this answers is not in the document]
Comment (stellarspot) [reply etcddb/etcddb_test.go]: removed write permission from the file.

File: etcddb/etcddb_server.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
+package etcddb
```
[the remark this answers is not in the document]
Comment (stellarspot) [reply etcddb/etcddb_server.go]: removed write permission from the file.

File: etcddb/etcddb_server.go [hunk 6]
```diff
@@ -76,10 +77,12 @@ func getEtcdConf(conf *PaymentChannelStorageServerConf, cluster string) *embed.C
...
-	fmt.Printf("name: '%s'\n", conf.ID)
-	fmt.Printf("client url: '%v'\n", clientURL)
-	fmt.Printf("peer   url: '%v'\n", peerURL)
-	fmt.Printf("initial cluster: '%s'\n", cluster)
+	log.WithFields(log.Fields{
+		"etcd id":         conf.ID,
+		"client url":      clientURL,
+		"peer   url":      peerURL,
+		"initial cluster": cluster,
+	}).Debug()
```
Comment (vsbogd) [inline etcddb/etcddb_server.go]: I would use Info() it is very useful to know which configuration values are actually used.

File: etcddb/etcddb_client.go [hunk 7]
```diff
@@ -6,15 +6,16 @@ import (
...
 
+	"github.com/spf13/viper"
 	"go.etcd.io/etcd/clientv3"
 )
 
 const (
-	// ConnectionTimeout connectio timeout
-	ConnectionTimeout = 5 * time.Second
+	// DefaultConnectionTimeout connection timeout in seconds
+	DefaultConnectionTimeout = 5
```
Comment (vsbogd) [inline etcddb/etcddb_client.go]: I am sorry for a late note but, it probably makes sense using milliseconds for timeout here.

**Pull request merged**